### PR TITLE
Enable amp-ad-exit to receive response from amp-analytics 3p vendor iframe

### PR DESCRIPTION
**Pull request by jonkeller**
Previous work implemented ability for amp-analytics to create a 3p vendor iframe.
Parallel PR #11306 enables the iframe to send a response to the pub page
This PR enables amp-ad-exit to receive those responses and substitute them into target URLs.

### CODE REVIEW

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 1]
```diff
@@ -100,18 +119,67 @@ export class AmpAdExit extends AMP.BaseElement {
...
-      for (const customVar in target.vars) {
-        if (customVar[0] == '_') {
-          vars[customVar] = () =>
-              args[customVar] == undefined ?
-                target.vars[customVar].defaultValue : args[customVar];
-          whitelist[customVar] = true;
+      for (const customVarName in target.vars) {
+        if (customVarName[0] == '_') {
+          const customVar =
+              /** @type {./config.Variable} */ (target.vars[customVarName]);
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: `!./config.Variable`

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 2]
```diff
@@ -100,18 +119,67 @@ export class AmpAdExit extends AMP.BaseElement {
...
+               The cross-domain iframe of vendorXYZ has sent the
+               following response for the creative:
+                 { priority: medium, category: W }
+               This is just example data. The keys/values in that object can
+               be any strings.
+               The code below will create substitutionFunctions['_pty'],
+               which in this example will return "medium".
+             */
+            substitutionFunctions[customVarName] = () => {
+              if (customVar.hasOwnProperty('iframeTransportSignal')) {
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Where does `customVar` come from? If `customVar` is an untrusted object, then this needs to be `Object.prototype.hasOwnProperty.call(customVar, 'iframeTransportSignal')`. On the other hand, if it came out of `JSON.parse`, then you should just do `'iframeTransportSignal' in customVar`.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 3]
```diff
@@ -100,18 +119,67 @@ export class AmpAdExit extends AMP.BaseElement {
...
+                  /* The vendor (in the example above, "vendorXYZ") has
+                     responded to this creative. Need to check whether that
+                     response contains a property that matches the
+                     second parameter to IFRAME_TRANSPORT_SIGNAL (ex:
+                     "priority") for this custom variable. If so, return the
+                     value in the response object that is associated with
+                     that key.
+                  */
+                  const responseKey = matches[2];
+                  if (vendorResponses.hasOwnProperty(responseKey)) {
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: See above.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 4]
```diff
@@ -100,18 +119,67 @@ export class AmpAdExit extends AMP.BaseElement {
...
+                    return vendorResponses[responseKey];
+                  }
+                }
+              }
+              // Either it's not a 3p analytics variable, or it is one but
+              // no matching response has been received yet.
+              return (args && args.hasOwnProperty(customVarName)) ?
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: See above.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 5]
```diff
@@ -190,6 +257,36 @@ export class AmpAdExit extends AMP.BaseElement {
...
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      let ampAdResourceId;
+      try {
+        ampAdResourceId = this.element.ownerDocument.defaultView
+          .frameElement.parentElement.getResourceId();
+      } catch (e) {
+        this.user().error(TAG,
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: I've never seen `this.user()` before. I should probably ask Hongfei or somebody from AMP Core what it's for...

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 6]
```diff
@@ -190,6 +257,36 @@ export class AmpAdExit extends AMP.BaseElement {
...
+        !responseMessage['creativeId'] ||
+        responseMessage['creativeId'] != ampAdResourceId) {
+        return;
+      }
+
+      dev().assert(responseMessage && responseMessage['message'],
+          'Received empty response from 3p analytics frame');
+      dev().assert(responseMessage && responseMessage['vendor'],
+          'Received response from 3p analytics frame that does not indicate' +
+        ' vendor');
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Indent?

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 7]
```diff
@@ -386,8 +408,7 @@ describes.realWin('amp-ad-exit', {
     if (!win.navigator) {
       win.navigator = {sendBeacon: () => false};
     }
-    const sendBeacon =
-        sandbox.stub(win.navigator, 'sendBeacon', () => true);
+    const sendBeacon = sandbox.stub(win.navigator, 'sendBeacon', () => true);
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: You can use `.returns(true)` here.

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 8]
```diff
@@ -522,5 +543,51 @@ describes.realWin('amp-ad-exit', {
     expect(open).to.have.been.calledWith(
         EXIT_CONFIG.targets.borderProtection.finalUrl, '_blank');
   });
+
+  it('should replace custom URL variables with 3P Analytics defaults', () => {
+    const open = sandbox.stub(win, 'open', () => {
```
Comment (taymonbeal) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: See above.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 2 above)
Comment (lannka) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: url-replacement-impl already supports function var.
you should be able to introduce new function by doing 
```
Services.urlReplacementsForDoc(this.getAmpDoc()).getVariableSource().setAsync('IFRAME_TRANSPORT_SIGNAL', resolver);
```

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Thanks, @lannka. I remember you said there was a better way, and was planning to ask for details but you beat me to it :)

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: It stems from the parseJson() call on line 239. Will change to 'x in y' form. Thanks.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 5 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: I think this was just a careless mistake. Removing.

Comment (taymonbeal) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: No, it's used elsewhere in this file.

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Hmm...I do see lots of occurrences: https://github.com/ampproject/amphtml/search?utf8=%E2%9C%93&q=%22this.user%28%29.error%22&type=

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 2 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Fixed.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 5 above)
Comment (zhouyx) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: We migrated to `this.user().error(` which equals to `user(this.element).error(`. With the change we can tell if the user error is from the parent AMP doc or from the friendly iframe.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 9]
```diff
@@ -187,9 +249,39 @@ export class AmpAdExit extends AMP.BaseElement {
...
+      user().error(TAG, 'Invalid JSON config', e);
       throw e;
     }
+
+    this.unlisten_ = listen(this.getAmpDoc().win, 'message', event => {
+      const responseMessage = deserializeMessage(getData(event));
+
+      let ampAdResourceId;
+      try {
+        ampAdResourceId = this.element.ownerDocument.defaultView
```
Comment (zhouyx) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: cache the ampAdResourceId in buildCallback?

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 10]
```diff
@@ -187,9 +249,39 @@ export class AmpAdExit extends AMP.BaseElement {
...
+        user().error(TAG,
+            'No friendly parent amp-ad element was found for amp-ad-exit tag.');
+      }
+
+      if (!responseMessage || !responseMessage['type'] ||
+        responseMessage['type'] != MessageType.IFRAME_TRANSPORT_RESPONSE ||
+        !responseMessage['creativeId'] ||
+        responseMessage['creativeId'] != ampAdResourceId) {
+        return;
+      }
```
Comment (zhouyx) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: let's create a helper method to validate incoming message

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 9 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Fixed.
Agree re: splitting, will do that tomorrow.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 10 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Fixed.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 11]
```diff
@@ -187,7 +250,66 @@ export class AmpAdExit extends AMP.BaseElement {
...
+        'vendor missing');
+    assertOriginMatchesVendor(expectedVendor, responseMessage['vendor']);
+  }
+
+  /**
+   * Gets the resource ID of the amp-ad element containing this AmpAdExit
+   * instance.
+   * @return {string}
+   */
+  getAmpAdResourceId() {
```
Comment (jonkeller) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: This is essentially the same as in amp-analytics, only the comments and error message text differ.
Any suggestions for a common place to put this? src/dom?

**Comment by jonkeller** [comment]
Closing this PR, as I have split it into #11461 and #11462.

**Pull request closed without merging**